src/datasets/utils/remote_utils.py

```diff
@@ -0,0 +1,198 @@
+            proxies (dict): Parameter passed to `requests.request` with the protocol to the URL of the proxy.
+            headers (dict, optional): Parameter passed to `requests.request` with the HTTP Headers to be sent with the
+                Request.
+            cookies (dict, optional): Parameter passed to `requests.request` with the cookies to be sent with the
+                Request.
+            allow_redirects (bool, default=True): Parameter passed to `requests.request` to enable/disable
+                redirections.
+            timeout (float, default=10): Parameter passed to `requests.request` to indicate how many seconds to wait
+                for the server to send data before giving up.
+            max_retries (int, default=0): Maximum number of retries, defaults to 0 (no retries).
+
+        Returns:
+            requests.Response
+        """
+        headers = copy.deepcopy(headers) or {}
+        headers["user-agent"] = cls.get_datasets_user_agent(user_agent=headers.get("user-agent"))
+        response = cls.request_with_retry(
+            verb="HEAD",
+            url=url,
+            proxies=proxies,
+            headers=headers,
+            cookies=cookies,
+            allow_redirects=allow_redirects,
+            timeout=timeout,
+            max_retries=max_retries,
+        )
+        return response
+
+    @classmethod
+    def get(cls, url, proxies=None, resume_size=0, headers=None, cookies=None, max_retries=0, callback=None):
+        """Wrapper around `requests` GET.
+
+        Args:
+            url (str): The URL of the resource to fetch.
+            proxies (dict): Parameter passed to `requests.request` with the protocol to the URL of the proxy.
+            resume_size (int): Part of the resource to be returned, passed to Range HTTP request Header as bytes range
+                start.
+            headers (dict, optional): Parameter passed to `requests.request` with the HTTP Headers to be sent with the
+                Request.
+            cookies (dict, optional): Parameter passed to `requests.request` with the cookies to be sent with the
+                Request.
+            max_retries (int, default=0): Maximum number of retries, defaults to 0 (no retries).
+            callback (callable, optional): Callback function to be called with each response data chunk.
+
+        Returns:
+            requests.Response or type returned by `callback`
+        """
+        headers = copy.deepcopy(headers) or {}
+        headers["user-agent"] = cls.get_datasets_user_agent(user_agent=headers.get("user-agent"))
+        if resume_size > 0:
+            headers["Range"] = "bytes=%d-" % (resume_size,)
+        with cls.request_with_retry(
+            verb="GET",
+            url=url,
+            stream=True,
+            proxies=proxies,
+            headers=headers,
+            cookies=cookies,
+            max_retries=max_retries,
+        ) as response:
```

Comment:
Do you really need to use a `with` statement here ?

Reply:
That is the key point! I use the `with` statement to make explicit that the connection is NOT closed (after calling get) when you pass the parameter `stream=True`.

Reply:
The explanation is here: https://requests.readthedocs.io/en/master/user/advanced/#body-content-workflow

> If you set `stream` to True when making a request, Requests cannot release the connection back to the pool unless you consume all the data or call Response.close. This can lead to inefficiency with connections. If you find yourself partially reading request bodies (or not reading them at all) while using `stream=True`, you should make the request within a `with` statement to ensure it's always closed:

    with requests.get('https://httpbin.org/get', stream=True) as r:
        # Do things with the response here.

Reply:
But `request_with_retry` is not a contextmanager right ? So using `with` doesn't change anything ? Or did I misunderstand something ?

Reply:
`request_with_retry` returns a `requests.Response` and this can be used as a context manager. When using it or not as a context manager? It depends whether the request connection has been kept open or has been closed. When you pass the parameter `stream=True`, the connection keeps open.